### PR TITLE
Core/Player: Interrupt flag capturing even if damage is absorbed.

### LINKED ISSUE #9270: Absorb Shields Stopping Interrupts While Flag Capping
### What client do you play on?

enGB

### Faction

- [X] Alliance
- [X] Horde

### Content Phase:

- [X] Generic
- [ ] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [ ] 50-59

### Current Behaviour
Original report: https://github.com/chromiecraft/chromiecraft/issues/1726

Absorbed damage doesn't stop flag capping in Arathi Basin (PVP)

Tester: Confirmed. Tested with priest and mage absorb shields with a warr hitting both, and neither was interrupted by the attacks.

https://user-images.githubusercontent.com/81782124/142760394-f6d218ad-c73d-45a6-bdda-98820188d5a6.mp4

Reported by Trinity as well: https://github.com/TrinityCore/TrinityCore/issues/6222 - a fix is also described there.

### Expected Blizzlike Behaviour

Taking direct damage, even if absorbed by a priest shield for example, should stop capping a flag.
It makes priest very strong at capping...

### Source

>By Lodum on 2010/04/25 at 9:41 AM (Patch 3.3.3) 
Even while shielded, they still get interrupted when dealing with the flag, afaik. At least, I always seem to. - - https://www.wowhead.com/forums/topic/disc-priest-pvp-how-to-interrupt-a-flag-capper-153563#p2008269

From the same thread (since deleted so needed to use an old screenshot):

76090136-52268280-5fbb-11ea-9c01-fe854a752c15

https://youtu.be/ZYb5SVo4dz0?t=75 

### Steps to reproduce the problem

Enter Arathi, pop a shield of some kind, and attempt to cap a flag while a second player applies direct (not DoT) damage. 

### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/7e2e6f8ee85f82f57a3c6e0f6f645b4c42292e7a

### Operating system

Ubuntu 20.04

### Modules

- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-chromie-xp](https://github.com/azerothcore/mod-chromie-xp)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-ip-tracker](https://github.com/azerothcore/mod-ip-tracker)
- [mod-low-level-arena](https://github.com/azerothcore/mod-low-level-arena)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-pvpstats-announcer](https://github.com/azerothcore/mod-pvpstats-announcer)
- [mod-queue-list-cache](https://github.com/azerothcore/mod-queue-list-cache)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [lua-carbon-copy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-custom-worldboss](https://github.com/55Honey/Acore_CustomWorldboss)
- [lua-level-up-reward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-recruit-a-friend](https://github.com/55Honey/Acore_RecruitAFriend)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)
- [lua-temp-announcements](https://github.com/55Honey/Acore_TempAnnouncements)
- [lua-zonecheck](https://github.com/55Honey/acore_Zonecheck)

### Customizations

None

### Server

ChromieCraft

### PR DESCRIPTION
**Pull request by Branel**
## Changes Proposed:
-  Flag capturing will be cancelled even if damage is absorbed.

## Issues Addressed:

- Closes #9270
- Closes https://github.com/chromiecraft/chromiecraft/issues/1726

## SOURCE:

Original report.
Thanks to the author and the contributor to TrinityCore that originally wrote this fix: @zwerg
## Tests Performed:

- Tested on my local build on Windows 10, I made two toons and applied PW:S and attempted to capture the flag while dealing damage on another character.

## How to Test the Changes:

 Create two opposite faction toons on your local server, one should be for instance a priest.
 make sure to:
 .learn all my class
 .debug bg 
queue Arathi Basin

1. Apply PW:S on yourself as the priest and capture any flag
2. Deal damage to the priest on the other toon by applying a DOT or a direct damage spell
3. See how the flag capturing has been cancelled although the damage has been absorbed.

## Known Issues and TODO List:

- I have only tested PW:S absorb so far against DOTS and direct damage. There could be more instances where flag capturing has incorrect behavior.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Entities/Unit/Unit.cpp [hunk 1]
```diff
@@ -822,6 +822,17 @@ uint32 Unit::DealDamage(Unit* attacker, Unit* victim, uint32 damage, CleanDamage
...
+        if (!damage && damagetype != DOT && cleanDamage && cleanDamage->absorbed_damage)
+            if (victim != attacker && victim->GetTypeId() == TYPEID_PLAYER)
+                if (Spell* spell = victim->m_currentSpells[CURRENT_GENERIC_SPELL])
+                    if (spell->getState() == SPELL_STATE_PREPARING)
+                    {
+                        uint32 interruptFlags = spell->m_spellInfo->InterruptFlags;
+                        if (interruptFlags & SPELL_INTERRUPT_FLAG_ABORT_ON_DMG)
+                            victim->InterruptNonMeleeSpells(false);
+                    }
+
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Unit/Unit.cpp:835]: brackets

Comment (Branel) [reply src/server/game/Entities/Unit/Unit.cpp:835]: Sorry been a while, is it correct now?

File: src/server/game/Entities/Unit/Unit.cpp [hunk 2]
```diff
@@ -822,6 +822,25 @@ uint32 Unit::DealDamage(Unit* attacker, Unit* victim, uint32 damage, CleanDamage
         else
             victim->RemoveAurasWithInterruptFlags(AURA_INTERRUPT_FLAG_TAKE_DAMAGE, 0);
 
+       // interrupt spells with SPELL_INTERRUPT_FLAG_ABORT_ON_DMG on absorbed damage (no dots)
+        if (!damage && damagetype != DOT && cleanDamage && cleanDamage->absorbed_damage)
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Unit/Unit.cpp:826]:
```suggestion
        // interrupt spells with SPELL_INTERRUPT_FLAG_ABORT_ON_DMG on absorbed damage (no dots)
        if (!damage && damagetype != DOT && cleanDamage && cleanDamage->absorbed_damage)
```

File: src/server/game/Entities/Unit/Unit.cpp (hunk 1 above)
Comment (Winfidonarleyan) [reply src/server/game/Entities/Unit/Unit.cpp]: Yes, good

File: src/server/game/Entities/Unit/Unit.cpp [hunk 3]
```diff
@@ -822,6 +822,25 @@ uint32 Unit::DealDamage(Unit* attacker, Unit* victim, uint32 damage, CleanDamage
...
+            if (victim != attacker && victim->GetTypeId() == TYPEID_PLAYER)
+            {
+                if (Spell* spell = victim->m_currentSpells[CURRENT_GENERIC_SPELL])
+                {
+                    if (spell->getState() == SPELL_STATE_PREPARING)
+                    {
+                        uint32 interruptFlags = spell->m_spellInfo->InterruptFlags;
+                        {
+                        if (interruptFlags & SPELL_INTERRUPT_FLAG_ABORT_ON_DMG)
+                            victim->InterruptNonMeleeSpells(false);
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Unit/Unit.cpp:837]:
```suggestion
                            if (interruptFlags & SPELL_INTERRUPT_FLAG_ABORT_ON_DMG)
                                victim->InterruptNonMeleeSpells(false);
```

File: src/server/game/Entities/Unit/Unit.cpp [hunk 4]
```diff
@@ -822,6 +822,25 @@ uint32 Unit::DealDamage(Unit* attacker, Unit* victim, uint32 damage, CleanDamage
...
+            {
+                if (Spell* spell = victim->m_currentSpells[CURRENT_GENERIC_SPELL])
+                {
+                    if (spell->getState() == SPELL_STATE_PREPARING)
+                    {
+                        uint32 interruptFlags = spell->m_spellInfo->InterruptFlags;
+                        {
+                            if (interruptFlags & SPELL_INTERRUPT_FLAG_ABORT_ON_DMG)
+                                victim->InterruptNonMeleeSpells(false);
+                        }
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Unit/Unit.cpp:838]:
```suggestion
                        uint32 interruptFlags = spell->m_spellInfo->InterruptFlags;
                        if (interruptFlags & SPELL_INTERRUPT_FLAG_ABORT_ON_DMG)
                            victim->InterruptNonMeleeSpells(false);
```

File: src/server/game/Entities/Unit/Unit.cpp [hunk 5]
```diff
@@ -822,6 +822,23 @@ uint32 Unit::DealDamage(Unit* attacker, Unit* victim, uint32 damage, CleanDamage
...
+        {
+            if (victim != attacker && victim->GetTypeId() == TYPEID_PLAYER)
+            {
+                if (Spell* spell = victim->m_currentSpells[CURRENT_GENERIC_SPELL])
+                {
+                    if (spell->getState() == SPELL_STATE_PREPARING)
+                    {
+                        uint32 interruptFlags = spell->m_spellInfo->InterruptFlags;
+                        if (interruptFlags & SPELL_INTERRUPT_FLAG_ABORT_ON_DMG)
+                            victim->InterruptNonMeleeSpells(false);
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Unit/Unit.cpp:836]:
```suggestion
                        if (interruptFlags & SPELL_INTERRUPT_FLAG_ABORT_ON_DMG)
                        {
                            victim->InterruptNonMeleeSpells(false);
                        }
```

**Comment by Branel** [comment]
Thanks 🙌

**Comment by temperrr** [comment]
Thanks for your PR!

I have tested the following

- [x] PWS and hitting flag capturer -> flag captureing gets interrupted
- [x] Divine Shield (paladin) and getting hit -> flag captureing gets interrupted
- [x] PWS and getting hit by a dot -> flag captureing gets not interrupted
- [x] No errors found during compiling

Looks good

**Pull request merged**